packages/indexer-agent/src/query-fees/allocations.ts

```diff
@@ -300,38 +316,88 @@ export class AllocationReceiptCollector implements ReceiptCollector {
+      const fee = BigNumber.from(receipt.fees).toHexString()
+      const feePadding = 33 - fee.length / 2
+      encodedReceipts.writeZeroes(feePadding)
+      encodedReceipts.writeHex(fee)
+      encodedReceipts.writeHex(receipt.id)
+      encodedReceipts.writeHex(receipt.signature)
+    }
+    return encodedReceipts
+  }
+
   private async obtainReceiptsVoucher(
     receipts: AllocationReceipt[],
   ): Promise<void> {
     const logger = this.logger.child({
       allocation: receipts[0].allocation,
     })
-
+    // Gross underestimated number of receipts the gateway take at once
+    const receipts_threshold = 25_000
+    let response
     try {
       logger.info(`Collect receipts for allocation`, {
         receipts: receipts.length,
       })
 
-      // Encode the receipt batch to a buffer
-      // [allocationId, receipts[]] (in bytes)
-      const encodedReceipts = new BytesWriter(20 + receipts.length * 112)
-      encodedReceipts.writeHex(receipts[0].allocation)
-      for (const receipt of receipts) {
-        // [fee, id, signature]
-        const fee = BigNumber.from(receipt.fees).toHexString()
-        const feePadding = 33 - fee.length / 2
-        encodedReceipts.writeZeroes(feePadding)
-        encodedReceipts.writeHex(fee)
-        encodedReceipts.writeHex(receipt.id)
-        encodedReceipts.writeHex(receipt.signature)
+      // All receipts can fit the gateway, make a single-shot collection
+      if (receipts.length <= receipts_threshold) {
+        const encodedReceipts = this.encodeReceiptBatch(receipts)
+
+        // Exchange the receipts for a voucher signed by the counterparty (aka the client)
+        response = await axios.post(
+          this.collectEndpoint.toString(),
+          encodedReceipts.unwrap().buffer,
+          { headers: { 'Content-Type': 'application/octet-stream' } },
+        )
+      } else {
+        // Split receipts in batches and collect partial vouchers
+        const partial_vouchers: Array<PartialVoucher> = []
+        for (let i = 0; i < receipts.length; i += receipts_threshold) {
+          const partial_receipts = receipts.slice(
+            i,
+            Math.min(i + receipts_threshold, receipts.length),
+          )
+          const encodedReceipts = this.encodeReceiptBatch(partial_receipts)
+
+          // Exchange the receipts for a partial voucher signed by the counterparty (aka the client)
+          response = await axios.post(
+            this.partialVoucherEndpoint.toString(),
+            encodedReceipts.unwrap().buffer,
+            { headers: { 'Content-Type': 'application/octet-stream' } },
+          )
+          const partial_voucher = response.data as PartialVoucher
+          partial_vouchers.push(partial_voucher)
+        }
+
+        // Take the partial vouchers and request for a full voucher
+        // (not sure if I can stright up pass the array in, but I want to try encoding them first
+        const voucherRequest = new BytesWriter(
```

Comment:
Let's also move this bytes writer to its own function similar to `encodeReceiptBatch()`?